### PR TITLE
Add field keywords to rewrite middleware

### PR DESCRIPTION
**Pull request by bwaz**
Add field keywords to the rewrite middleware. First handled keyword will be edns0. Without a keyword used the NewSimpleRule Rule will be used. Logic for handling edns0 will come soon. 

We can add "simple" as well to be backwards compatible.

### CODE REVIEW

File: middleware/rewrite/field.go [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+
+	"github.com/miekg/dns"
+)
+
+// Add more fields keywords here
+var Fields = map[string]int{
+	"edns0": 1,
+}
+
+// NewFieldRule creates a new Field Rule
```
Comment (miekg) [inline middleware/rewrite/field.go]: Nit: needs closing dot.

File: middleware/rewrite/field.go [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+func NewFieldRule(field string, args []string) Rule {
+	switch field {
```
Comment (miekg) [inline middleware/rewrite/field.go]: bogus function body because WIP?

File: middleware/rewrite/setup.go [hunk 3]
```diff
@@ -108,11 +108,19 @@ func rewriteParse(c *caddy.Controller) ([]Rule, error) {
...
+			rule = NewRule(args)
 			simpleRules = append(simpleRules, rule)
 		}
 	}
 
 	// put simple rules in front to avoid regexp computation for them
 	return append(simpleRules, regexpRules...), nil
 }
+
+func NewRule(args []string) Rule {
```
Comment (miekg) [inline middleware/rewrite/setup.go]: Docs

**Review by miekg — COMMENTED** [review]
I agree with the direction we're going here. The only thing I would ask is to be mindful of the language:
1. extensibility  
2. uniformity with the rest of coredns

**Comment by bwaz** [comment]
@miekg  @johnbelamaric  
Init work, edns0 or any other keywords can be added in a new PR
This change makes field a requirement.

File: middleware/rewrite/README.md [hunk 4]
```diff
@@ -7,24 +7,29 @@ accommodate most dynamic back-end applications.
...
 ~~~
 
+* **FIELD** is (`qtype`, `class`, `qname`, ...) 
 * **FROM** is the exact name of type to match
 * **TO** is the destination name or type to rewrite to
 
-If from *and* to look like a DNS type (`A`, `MX`, etc.), the type of the message will be rewriten;
-e.g., to rewrite ANY queries to HINFO, use `rewrite ANY HINFO`.
+When the FIELD is `qtype` and FROM is (`A`, `MX`, etc.), the type of the message will be rewriten;
+e.g., to rewrite ANY queries to HINFO, use `rewrite qtype ANY HINFO`.
```
Comment (miekg) [inline middleware/rewrite/README.md]: think we should rename qtype to just `type`, we also just use `class`

File: middleware/rewrite/README.md [hunk 5]
```diff
@@ -7,24 +7,29 @@ accommodate most dynamic back-end applications.
...
 
-If it does not look like a type a name is assumed and the qname in the message is rewritten; this
-needs to be a full match of the name, e.g., `rewrite miek.nl example.org`.
+When the FIELD is `qname` the qname in the message is rewritten; this
+needs to be a full match of the name, e.g., `rewrite qname miek.nl example.org`.
 
 If you specify multiple rules and an incoming query matches on multiple (simple) rules, only
 the first rewrite is applied.
 
+Additional FIELD keywords may be implemented to support more rewrite use-cases.
```
Comment (miekg) [inline middleware/rewrite/README.md]: This more of a package level documention then something for a readme. Think this readme's should be focused on usage

File: middleware/rewrite/rewrite.go [hunk 6]
```diff
@@ -59,64 +56,6 @@ func (rw Rewrite) Name() string { return "rewrite" }
...
-			return RewriteDone
-		}
-	}
-
-	if s.From == r.Question[0].Name {
-		r.Question[0].Name = s.To
-		return RewriteDone
-	}
-	return RewriteIgnored
+	// init
```
Comment (miekg) [inline middleware/rewrite/rewrite.go]: init -> `New returns a new rule.`

**Review by miekg — COMMENTED** [review]
lgtm - modulo some nits. I'll update the 005 release wiki

File: middleware/rewrite/README.md (hunk 4 above)
Comment (miekg) [reply middleware/rewrite/README.md]: and type as well?

File: middleware/rewrite/rewrite_test.go [hunk 7]
```diff
@@ -20,9 +20,9 @@ func TestRewrite(t *testing.T) {
 	rw := Rewrite{
 		Next: middleware.HandlerFunc(msgPrinter),
 		Rules: []Rule{
-			NewSimpleRule("from.nl.", "to.nl."),
-			NewSimpleRule("CH", "IN"),
-			NewSimpleRule("ANY", "HINFO"),
+			Fields["qname"].New("from.nl.", "to.nl."),
```
Comment (miekg) [inline middleware/rewrite/rewrite_test.go]: We should try to use `example.*` zone in these tests (yes, there is a of `miek.nl` and `.nl` out there still)

File: middleware/rewrite/README.md (hunk 4 above)
Comment (bwaz) [reply middleware/rewrite/README.md]: I'll change it to class, type, name

File: middleware/rewrite/rewrite_test.go (hunk 7 above)
Comment (bwaz) [reply middleware/rewrite/rewrite_test.go]: I'm having an issue with testing *
```
diff --git a/middleware/rewrite/rewrite_test.go b/middleware/rewrite/rewrite_test.go
index ffb1b4a..63d2f84 100644
--- a/middleware/rewrite/rewrite_test.go
+++ b/middleware/rewrite/rewrite_test.go
@@ -20,9 +20,10 @@ func TestRewrite(t *testing.T) {
        rw := Rewrite{
                Next: middleware.HandlerFunc(msgPrinter),
                Rules: []Rule{
-                       Fields["qname"].New("from.nl.", "to.nl."),
+                       Fields["name"].New("from.nl.", "to.nl."),
+                       Fields["name"].New("myexample.*", "myexample.com."),
                        Fields["class"].New("CH", "IN"),
-                       Fields["qtype"].New("ANY", "HINFO"),
+                       Fields["type"].New("ANY", "HINFO"),
                },
                noRevert: true,
        }
@@ -35,6 +36,7 @@ func TestRewrite(t *testing.T) {
                toT   uint16
                toC   uint16
        }{
+               {"myexample.org.", dns.TypeA, dns.ClassINET, "myexample.com.", dns.TypeA, dns.ClassINET},
                {"from.nl.", dns.TypeA, dns.ClassINET, "to.nl.", dns.TypeA, dns.ClassINET},
                {"a.nl.", dns.TypeA, dns.ClassINET, "a.nl.", dns.TypeA, dns.ClassINET},
                {"a.nl.", dns.TypeA, dns.ClassCHAOS, "a.nl.", dns.TypeA, dns.ClassINET},

```

```
--- FAIL: TestRewrite (0.00s)
	rewrite_test.go:61: Test 0: Expected Name to be 'myexample.com.' but was 'myexample.org.'

```

Comment (bwaz) [reply middleware/rewrite/rewrite_test.go]: Anything I'm doing wrong with the setup? Can we defer to new issue if functionality is not implemented yet?

File: middleware/rewrite/README.md [hunk 8]
```diff
@@ -7,24 +7,26 @@ accommodate most dynamic back-end applications.
...
+rewrite FIELD FROM TO
 ~~~
 
+* **FIELD** is (`type`, `class`, `name`, ...)
 * **FROM** is the exact name of type to match
 * **TO** is the destination name or type to rewrite to
 
-If from *and* to look like a DNS type (`A`, `MX`, etc.), the type of the message will be rewriten;
-e.g., to rewrite ANY queries to HINFO, use `rewrite ANY HINFO`.
+When the FIELD is `type` and FROM is (`A`, `MX`, etc.), the type of the message will be rewriten;
```
Comment (johnbelamaric) [inline middleware/rewrite/README.md]: s/rewriten/rewritten/

File: middleware/rewrite/README.md [hunk 9]
```diff
@@ -7,24 +7,26 @@ accommodate most dynamic back-end applications.
...
+When the FIELD is `class` and FROM is (`IN`, `CH`, or `HS`) the class of the message will be
+rewritten; e.g., to rewrite CH queries to IN use `rewrite class CH IN`.
 
-If it does not look like a type a name is assumed and the qname in the message is rewritten; this
-needs to be a full match of the name, e.g., `rewrite miek.nl example.org`.
+When the FIELD is `name` the query name in the message is rewritten; this
+needs to be a full match of the name, e.g., `rewrite name miek.nl example.org`.
 
 If you specify multiple rules and an incoming query matches on multiple (simple) rules, only
 the first rewrite is applied.
```
Comment (johnbelamaric) [inline middleware/rewrite/README.md:27]: Is this true? Is this what we want? If you can specify multiple re-write middleware, then I assume they all act independently.

File: middleware/rewrite/setup.go [hunk 10]
```diff
@@ -108,8 +106,10 @@ func rewriteParse(c *caddy.Controller) ([]Rule, error) {
 
 		// the only unhandled case is 2 and above
 		default:
-			rule = NewSimpleRule(args[0], strings.Join(args[1:], " "))
-			simpleRules = append(simpleRules, rule)
+			if _, ok := Fields[args[0]]; ok {
+				rule = Fields[args[0]].New(args[1:]...)
+				simpleRules = append(simpleRules, rule)
+			}
```
Comment (johnbelamaric) [inline middleware/rewrite/setup.go]: else return an error so the user knows they made a typo...

**Review by johnbelamaric — CHANGES_REQUESTED** [review]
Looking good.

File: middleware/rewrite/README.md (hunk 9 above)
Comment (miekg) [reply middleware/rewrite/README.md:27]: It can be done, but it fairly useless, because clients reject responses that are fiddled with, also DNSSEC...

**Comment by miekg** [comment]
[ Quoting <notifications@github.com> in "Re: [miekg/coredns] Add field keywo..." ]
>bwaz commented on this pull request.
>
>
>
>> @@ -20,9 +20,9 @@ func TestRewrite(t *testing.T) {
> 	rw := Rewrite{
> 		Next: middleware.HandlerFunc(msgPrinter),
> 		Rules: []Rule{
>-			NewSimpleRule("from.nl.", "to.nl."),
>-			NewSimpleRule("CH", "IN"),
>-			NewSimpleRule("ANY", "HINFO"),
>+			Fields["qname"].New("from.nl.", "to.nl."),
>
>I'm having an issue with testing *
>```
>diff --git a/middleware/rewrite/rewrite_test.go b/middleware/rewrite/rewrite_test.go
>index ffb1b4a..63d2f84 100644
>--- a/middleware/rewrite/rewrite_test.go
>+++ b/middleware/rewrite/rewrite_test.go
>@@ -20,9 +20,10 @@ func TestRewrite(t *testing.T) {
>        rw := Rewrite{
>                Next: middleware.HandlerFunc(msgPrinter),
>                Rules: []Rule{
>-                       Fields["qname"].New("from.nl.", "to.nl."),
>+                       Fields["name"].New("from.nl.", "to.nl."),
>+                       Fields["name"].New("myexample.*", "myexample.com."),
>                        Fields["class"].New("CH", "IN"),
>-                       Fields["qtype"].New("ANY", "HINFO"),
>+                       Fields["type"].New("ANY", "HINFO"),
>                },
>                noRevert: true,
>        }
>@@ -35,6 +36,7 @@ func TestRewrite(t *testing.T) {
>                toT   uint16
>                toC   uint16
>        }{
>+               {"myexample.org.", dns.TypeA, dns.ClassINET, "myexample.com.", dns.TypeA, dns.ClassINET},
>                {"from.nl.", dns.TypeA, dns.ClassINET, "to.nl.", dns.TypeA, dns.ClassINET},
>                {"a.nl.", dns.TypeA, dns.ClassINET, "a.nl.", dns.TypeA, dns.ClassINET},
>                {"a.nl.", dns.TypeA, dns.ClassCHAOS, "a.nl.", dns.TypeA, dns.ClassINET},
>
>```
>
>```
>--- FAIL: TestRewrite (0.00s)
>	rewrite_test.go:61: Test 0: Expected Name to be 'myexample.com.' but was 'myexample.org.'
>
>```

Oh, yeah, this might not work, because you're doing multiple rules chained after
each other (unless you fixed that as well). At least in CoreDNS <= 004 this
chaining never worked.

**Comment by bwaz** [comment]
I suggest splitting any additional work from this point to new PR

**Comment by miekg** [comment]
[ Quoting <notifications@github.com> in "Re: [miekg/coredns] Add field keywo..." ]
>I suggest splitting any additional work from this point to new PR

ack.

**Review by johnbelamaric — APPROVED** [review]

**Comment by johnbelamaric** [comment]
lgtm

**Pull request merged**